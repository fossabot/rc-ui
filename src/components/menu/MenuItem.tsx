import React, { useCallback, useContext } from 'react';
import classNames from 'classnames';
import MenuContext from './MenuContext';

export interface MenuItemProps {
  key: string;
  disabled?: boolean;
  style?: React.CSSProperties;
  className?: string;
  rootPrefixCls?: string;
  eventKey?: string;
}

const MenuItem: React.FC<MenuItemProps> = props => {
  const { rootPrefixCls, className, style, disabled, children, eventKey } = props;
  const { onSelect, selectedKey } = useContext(MenuContext);
  const classes = classNames(`${rootPrefixCls}-item`, className, {
    [`${rootPrefixCls}-item-disabled`]: disabled,
    [`${rootPrefixCls}-item-selected`]: eventKey === selectedKey,
  });
  const handleClick = useCallback(
    (e: React.MouseEvent<HTMLLIElement>) => {
      const info = {
        key: eventKey as string,
        domEvent: e,
      };
      if (!disabled) {
        onSelect!(info);
      }
    },
    [eventKey, disabled, onSelect],
  );
  const onKeyDown = useCallback(
    (e: React.KeyboardEvent<HTMLElement>): boolean | undefined => {
      const { keyCode } = e;
      if (keyCode === 13 && !disabled) {
        onSelect!({ key: eventKey as string, domEvent: e });
        return true;
      }
      return undefined;
    },
    [eventKey, disabled, onSelect],
  );

  return (
    <li
      role="menuitem"
      className={classes}
      style={style}
      onClick={handleClick}
      onKeyDown={onKeyDown}
    >
      {children}
    </li>
  );
};

MenuItem.displayName = '__KS_MenuItem__';

export default MenuItem;
