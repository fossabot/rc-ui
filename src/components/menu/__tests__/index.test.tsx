import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import mountTest from '../../../tests/mountTest';
import Menu, { MenuProps } from '..';
import MenuItem from '../MenuItem';

const genMenu = (props?: MenuProps) => {
  return (
    <Menu defaultSelectedKey="1" {...props}>
      <MenuItem key="1">Active</MenuItem>
      <MenuItem key="2">Normal</MenuItem>
      <MenuItem key="3" disabled>
        Disabled
      </MenuItem>
    </Menu>
  );
};
const prefixCls = 'ks-menu';
const activeClass = `${prefixCls}-item-selected`;

describe('Menu', () => {
  it('renders correctly', () => {
    expect(genMenu()).toMatchSnapshot();
  });

  mountTest(() => genMenu());

  it('should render correct Menu and MenuItem based on the specified props', () => {
    const wrapper = render(genMenu());
    const menuElement = wrapper.getByRole('menu');
    const activeElement = wrapper.getByText('Active');
    const disabledElement = wrapper.getByText('Disabled');

    expect(menuElement).toBeInTheDocument();
    expect(menuElement).toHaveClass(`${prefixCls} ${prefixCls}-horizontal`);
    expect(activeElement).toHaveClass(activeClass);
    expect(disabledElement).toHaveClass(`${prefixCls}-item-disabled`);
    expect(menuElement.getElementsByTagName('li').length).toEqual(3);
    expect(menuElement.getElementsByTagName('li')[0]).toHaveClass(`${prefixCls}-item`);
  });

  it('should render vertical mode when mode is set to vertical', () => {
    const wrapper = render(genMenu({ mode: 'vertical' }));
    const menuElement = wrapper.getByRole('menu');

    expect(menuElement).toBeInTheDocument();
    expect(menuElement).toHaveClass(`${prefixCls} ${prefixCls}-vertical`);
  });

  it('click items should change active and call the right callback', () => {
    const testProps = { onSelect: jest.fn() };
    const wrapper = render(genMenu(testProps));
    const activeElement = wrapper.getByText('Active');
    const disabledElement = wrapper.getByText('Disabled');
    const normalItem = wrapper.getByText('Normal');

    fireEvent.click(normalItem);
    expect(normalItem).toHaveClass(activeClass);
    expect(activeElement).not.toHaveClass(activeClass);
    expect(testProps.onSelect).toHaveBeenCalledWith(expect.any(Object));
    fireEvent.click(disabledElement);
    expect(normalItem).toHaveClass(activeClass);
    expect(disabledElement).not.toHaveClass(activeClass);
  });
});
