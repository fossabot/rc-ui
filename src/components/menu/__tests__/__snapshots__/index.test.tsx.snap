// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `
<Menu
  defaultSelectedKey="1"
  mode="horizontal"
  onSelect={[Function]}
>
  <__KS_MenuItem__>
    Active
  </__KS_MenuItem__>
  <__KS_MenuItem__>
    Normal
  </__KS_MenuItem__>
  <__KS_MenuItem__
    disabled={true}
  >
    Disabled
  </__KS_MenuItem__>
  <__KS_SubMenu__
    title="SubItems"
  >
    <__KS_MenuItem__>
      First SubItem
    </__KS_MenuItem__>
    <__KS_MenuItem__>
      Second SubItem
    </__KS_MenuItem__>
  </__KS_SubMenu__>
</Menu>
`;
